File: Python/Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from operator import attrgetter
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        sortedList = []
        for l in lists:
            cur = l
            while cur:
                sortedList.append(cur)
                cur = cur.next
        sortedList = sorted(sortedList, key = attrgetter('val'))
        
        for i in range(len(sortedList)-1):
            sortedList[i].next = sortedList[i+1]
        if len(sortedList) > 0:
            sortedList[len(sortedList)-1].next = None
        
        if sortedList:
            return sortedList[0]
        else:
            return None
            

File: Java/src/Merge k Sorted Lists.java
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) { val = x; }
 * }
 */
public class Solution
{
  public ListNode mergeKLists(ListNode[] lists)
  {
    Queue<ListNode> queue = new PriorityQueue(new Comparator<ListNode>()
    {
      @Override
      public int compare(ListNode node1, ListNode node2)
      {
        return node1.val - node2.val;
      }
    });

    for (ListNode l : lists)
    {
      if (l != null)
        queue.add(l);
    }

    ListNode dummy = new ListNode(0);
    ListNode ret = dummy;
    while (!queue.isEmpty())
    {
      ret.next = queue.poll();
      ret = ret.next;
      if (ret.next != null)
      {
        queue.offer(ret.next);
      }
    }
    return dummy.next;

  }
}
